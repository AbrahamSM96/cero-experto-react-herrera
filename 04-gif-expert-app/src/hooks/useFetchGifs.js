

import { useState, useEffect } from 'react'

export const useFetchGifs = (category) => {

    const [images, setImages] = useState({
        data: [],
        loading: true
    })

    const getGifs = async (category) => {
        const URI = 'https://api.giphy.com/v1/gifs/'
        const API_KEY = '4h6xSD2bAOTCgvV0VEMDK091XK6maDTw'
        const CATEGORY = encodeURI(category)
        const LIMIT = 10
        const API = `${URI}search?q=${CATEGORY}&limit=${LIMIT}&api_key=${API_KEY}`
        const response = await fetch(API)
        const { data } = await response.json()
        const gifs = data.map((img) => {
            return {
                id: img.id,
                title: img.title,
                url: img.images?.downsized_medium.url
            }
        })
        setImages({
            data: gifs,
            loading: false
        })
    }

    useEffect(() => {
        getGifs(category)
    }, [category])

    return images
}