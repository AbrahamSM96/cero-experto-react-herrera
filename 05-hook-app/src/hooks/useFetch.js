import { useEffect, useState } from "react";

export const useFetch = (url) => {
  const [data, setData] = useState({
    data: null,
    loading: true,
    error: null,
  });

  useEffect(() => {
    setData({ data: null, loading: true, error: null });
    fetch(url)
      .then((resp) => resp.json())
      .then((data) => {
        setData({
          loading: false,
          error: null,
          data,
        });
      });
  }, [url]);

  return data;
};
