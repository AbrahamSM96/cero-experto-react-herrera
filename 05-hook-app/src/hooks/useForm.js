import { useState } from 'react';

export const useForm = (initialState = {}) => {
  const [value, setValue] = useState(initialState);

  const handleInputChange = ({ target }) => {
    const { name, value } = target;
    setValue({
      ...value,
      [name]: value,
    });
  };

  const onResetForm = () => {
    setValue(initialState);
  };

  return { ...value, value, handleInputChange, onResetForm };
};
