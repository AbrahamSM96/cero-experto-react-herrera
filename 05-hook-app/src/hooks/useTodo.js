import { useEffect, useReducer } from 'react';
import { todoReducer } from '../components/08-useReducer/todoReducer';

export const useTodo = () => {
  const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];
  };
  const initialState = [];

  const [todos, dispatch] = useReducer(todoReducer, initialState, init);
  const TODO_COUNT = todos.length;
  const TODO_PENDING = todos.filter((todo) => !todo.done).length;

  const handleNewTodo = (todo) => {
    const action = {
      type: '[TODO] Add Todo',
      payload: todo,
    };
    dispatch(action);
  };

  const handleDeleteTodo = (id) => {
    dispatch({
      type: '[TODO] Remove Todo',
      payload: id,
    });
  };

  const handleToggleTodo = (id) => {
    dispatch({
      type: '[TODO] Toggle Todo',
      payload: id,
    });
    console.log(id);
  };

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);

  return {
    todos,
    handleNewTodo,
    handleDeleteTodo,
    handleToggleTodo,
    TODO_COUNT,
    TODO_PENDING,
  };
};
