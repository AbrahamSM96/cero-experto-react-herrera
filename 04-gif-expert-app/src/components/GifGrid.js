import React from "react";
import PropTypes from "prop-types";
import { useFetchGifs } from "../hooks/useFetchGifs";
import GifGridItem from "./GifGridItem";

export default function GifGrid({ category }) {
    const { data = [], loading } = useFetchGifs(category);

    return (
        <>
            <h3 className="animate__animated animate__fadeInUp">{category}</h3>

            <div className="cardGrid">
                {loading ? (
                    <h1 className="animate__animated animate__flash">Loading</h1>
                ) : (
                    data.map((item) => <GifGridItem key={`${item.id}`} {...item} />)
                )}
            </div>
        </>
    );
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired,
};
