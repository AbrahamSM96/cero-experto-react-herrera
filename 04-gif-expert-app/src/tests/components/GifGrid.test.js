import React from 'react'
import '@testing-library/jest-dom'
import { shallow } from 'enzyme';
import GifGrid from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs')

describe('test in <GifGrid></GifGrid>', () => {

    const CATEGORY = 'perro'
    const wrapper = shallow(<GifGrid category={CATEGORY} />)
    test('should show the component GifGrid', () => {
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        })
        expect(wrapper).toMatchSnapshot()
    });

    test('should show items when load the images with useFetchGifs', () => {

        const gifs = [{
            id: 'ABC',
            url: 'https://localhost/cualquiercosa.jpg',
            title: 'HOLA PERRO'
        }]

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        })

        expect(wrapper.find('h1').exists()).toBe(false)




    });
});
