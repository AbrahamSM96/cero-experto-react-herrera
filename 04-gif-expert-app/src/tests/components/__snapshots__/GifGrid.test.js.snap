// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test in <GifGrid></GifGrid> should show the component GifGrid 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeInUp"
  >
    perro
  </h3>
  <div
    className="cardGrid"
  >
    <h1
      className="animate__animated animate__flash"
    >
      Loading
    </h1>
  </div>
</Fragment>
`;
